Add PParamUpdates to the plutus context, by transforming them to Plutus Data

TimSheard opened the pull request:
Adds a new class ToPlutusData 
Uses table driven approach to make instance for PParamsUpdate

resolves #3994 

# Checklist

- [x] Commit sequence broadly makes sense and commits have useful messages
- [x] New tests are added if needed and existing tests are updated
- [ ] When applicable, versions are updated in `.cabal` and `CHANGELOG.md` files according to the
      [versioning process](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process).
- [ ] The version bounds in `.cabal` files for all affected packages are updated. **If you change the bounds in a cabal file, that package itself must have a version increase.** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process))
- [ ] All visible changes are prepended to the latest section of a `CHANGELOG.md` for the affected packages. **New section is never added with the code changes.** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#changelogmd))
- [x] Code is formatted with [`fourmolu`](https://github.com/fourmolu/fourmolu) (use `scripts/fourmolize.sh`)
- [x] Cabal files are formatted (use `scripts/cabal-format.sh`)
- [x] [`hie.yaml`](https://github.com/intersectmbo/cardano-ledger/blob/master/hie.yaml) has been updated (use `scripts/gen-hie.sh`)
- [x] Self-reviewed the diff

TimSheard committed:
Add Translation of PParamsUpdate to PlutusContext:

* Add `Cardano.Ledger.Core.Plutus.ToPlutusData`
* Add ToPlutusData class
* Add ToPlutusData instances in `Cardano.Ledger.Conway.PParams`
* Use table driven approach to make instance for PParamsUpdate
* Property tests added
* Add module `Cardano.Ledger.Conway.Plutus.Context`
* Add class `ToPlutusData (PParamsUpdate era) => ConwayEraPlutusTxInfo (l :: Language) era`
* Complete module `Cardano.Ledger.Conway.TxInfo` using `ConwayEraPlutusTxInfo` instance

TimSheard commented:
Still a work in progress as it doesn't do anything yet with the plutus context

lehins commented on libs/cardano-ledger-core/src/Cardano/Ledger/Plutus/ToPlutusData.hs:
```diff
@@ -0,0 +1,259 @@
+
+-- | Pair the tag, and exisitenially hide the type of the lens for the field with that Lens'
+data PParam era where
+  PParam :: (DecCBOR t, ToPlutusData t) => Integer -> Lens' (PParamsUpdate era) (StrictMaybe t) -> PParam era
```
Let's change this to `Word`, since that is what used for CBOR tags. Conversion to `Integer` is cheap for the ToPlutusData conversion, but doing other operations on Integer and conversion back to Word, are more expensive, which is unnecessary.

Also we are not doing any arithmetic on those tags, so there are no dangers of overflow
```suggestion
  PParam :: (DecCBOR t, ToPlutusData t) => Word -> Lens' (PParamsUpdate era) (StrictMaybe t) -> PParam era
```

lehins commented on libs/cardano-ledger-core/src/Cardano/Ledger/Plutus/ToPlutusData.hs:
```diff
@@ -0,0 +1,259 @@
+pparamUpdateToData pparamMap ppu = Map (concat (Map.mapWithKey toMapPair pparamMap))
+  where
+    toMapPair tag (PParam _ lense) =
+      case ppu ^. lense of
```
Convention is to use `L` suffix for lenses
```suggestion
    toMapPair tag (PParam _ ppuL) =
      case ppu ^. ppuL of
```

lehins commented on libs/cardano-ledger-core/src/Cardano/Ledger/Plutus/ToPlutusData.hs:
```diff
@@ -0,0 +1,259 @@
+
+-- | Translate a PParamUpdate to Data, using the PParamMap.
+pparamUpdateToData :: Map.Map Integer (PParam era) -> PParamsUpdate era -> Data
+pparamUpdateToData pparamMap ppu = Map (concat (Map.mapWithKey toMapPair pparamMap))
```
This operation is better done with a fold, which will also guarantee the correct order. (`toMapPair` will need to be adjusted to handle the acc)
```suggestion
pparamUpdateToData pparamMap ppu = Map (Map.foldrWithKey' toMapPair [] pparamMap)
```

lehins commented on libs/cardano-ledger-core/src/Cardano/Ledger/Plutus/ToPlutusData.hs:
```diff
@@ -0,0 +1,259 @@
+-- Tools for building ToPlutusData instances for (PParamUpdates era) for many eras.
+
+-- | Pair the tag, and exisitenially hide the type of the lens for the field with that Lens'
+data PParam era where
```
This type needs to go into `Cardano.Ledger.Core.PParams`, since it is no longer related to just the PlutusData. You'll need to invert the dependency of modules, which will work when my other suggestions are applied

TimSheard replied:
done

lehins commented on libs/cardano-ledger-core/src/Cardano/Ledger/Plutus/ToPlutusData.hs:
```diff
@@ -0,0 +1,259 @@
+            else ppu
+pparamUpdateFromData _ _ = Nothing
+
+-- ========================================================
```
This experiment with using this for pre-conway pparams for decoding shows that we can use it for things other than ToPlutusData. For this reason everything that is below should travel to `Cardano.Ledger.Shelley.PParams`

Although, please don't expand on it any further in this PR. We have a very high priority on the original issue of translating PParams to the Plutus context.

TimSheard replied:
So I removed all this.

lehins commented on eras/conway/impl/src/Cardano/Ledger/Conway/PParamsUpdate.hs:
```diff
@@ -0,0 +1,127 @@
+{-# LANGUAGE TypeApplications #-}
+{-# OPTIONS_GHC -Wno-orphans #-}
+
+module Cardano.Ledger.Conway.PParamsUpdate where
```
This is plutus specific and is not related in any way to updating plutus parameters. Let's create a module with this name instead. (And yes, please create a new subfolder)
```suggestion
module Cardano.Ledger.Conway.Plutus.ToPlutusData where
```

lehins commented on eras/conway/impl/src/Cardano/Ledger/Conway/PParamsUpdate.hs:
```diff
@@ -0,0 +1,127 @@
+      <*> fromPlutusData econ
+      <*> fromPlutusData tech
+      <*> fromPlutusData gov
+      <*> fromPlutusData treasury
```
Stick with non semantic names as you've done for `PoolVotingThresholds`, since the order is very important and it is easier to see the order is correct with something like letters of alphabet or with numerics (eg. `x00`, `x01`, ..) 
```suggestion
  fromPlutusData (List [a, b, c, d, e, f, g, h, i, j]) =
    DRepVotingThresholds
      <$> fromPlutusData a
      <*> fromPlutusData b
      <*> fromPlutusData c
      <*> fromPlutusData d
      <*> fromPlutusData e
      <*> fromPlutusData f
      <*> fromPlutusData g
      <*> fromPlutusData h
      <*> fromPlutusData i
      <*> fromPlutusData j
```

lehins commented on eras/conway/impl/src/Cardano/Ledger/Conway/PParamsUpdate.hs:
```diff
@@ -0,0 +1,127 @@
+-- | Instance for ShelleyEra
+instance Crypto c => ToPlutusData (PParamsUpdate (ShelleyEra c)) where
+  toPlutusData = pparamUpdateToData shelleyPParamMap
+  fromPlutusData = pparamUpdateFromData shelleyPParamMap
```
This must be removed. We do not want provide functionality for passing shelley pparams to plutus
```suggestion
```

lehins commented on libs/cardano-ledger-core/src/Cardano/Ledger/Plutus/ToPlutusData.hs:
```diff
@@ -0,0 +1,259 @@
+-- | Decode any era, given the right (Map.Map Integer (PParam era))
+decodePParamsUpdate :: EraPParams era => Map.Map Integer (PParam era) -> Decoder s (PParamsUpdate era)
+decodePParamsUpdate ppMap = do
+  mapParts <- decodeMapContents $ do
```
This is incorrect, because `decodeMapContents` does not guard against duplicates.

Can you please remove anything related to decoding form this PR. That is not the goal of this work and when we do get to using this approach for decoding, I would like to have it all contained in a separate PR.

TimSheard replied:
done

lehins commented on libs/cardano-ledger-core/src/Cardano/Ledger/Plutus/ToPlutusData.hs:
```diff
@@ -0,0 +1,259 @@
+    | all isJust mlist = Just $ catMaybes mlist
+    | otherwise = Nothing
+    where
+      mlist = map fromPlutusData xs
```
```suggestion
  fromPlutusData (List xs) = mapM fromPlutusData xs
```

lehins commented on libs/cardano-ledger-core/src/Cardano/Ledger/Plutus/ToPlutusData.hs:
```diff
@@ -0,0 +1,259 @@
+          x <- fromPlutusData d1
+          y <- fromPlutusData d2
+          Just (x, y)
+  fromPlutusData _ = Nothing
```
```suggestion
  fromPlutusData (Map pairs) =
    Map.fromList <$> 
    mapM (\(k, v) -> (,) <$> fromPlutusData k <*> fromPlutusData v)) pairs
```

lehins commented on libs/cardano-ledger-core/src/Cardano/Ledger/Plutus/ToPlutusData.hs:
```diff
@@ -0,0 +1,259 @@
+  fromPlutusData (Constr 0 []) = Just NeutralNonce
+  fromPlutusData (Constr 1 [B bytes]) = Nonce <$> (hashFromBytes bytes)
+  fromPlutusData _ = Nothing
+
```
We are never translating this to plutus. So, this instance should be removed
```suggestion
```

lehins commented on libs/cardano-ledger-core/src/Cardano/Ledger/Plutus/ToPlutusData.hs:
```diff
@@ -0,0 +1,259 @@
+
+instance ToPlutusData Coin where
+  toPlutusData (Coin n) = I n
+  fromPlutusData (I n) | n > 0 = Just (Coin n) -- DO WE NEED TO TEST n > 0 ?
```
We don't. Unfortunately there is no invariant on the Coin type that it can never be negative. In fact there are a couple of places in the code where it can be negative, albeit for invalid transactions.
```suggestion
  fromPlutusData (I n) = Just (Coin n)
```

lehins commented on libs/cardano-ledger-core/src/Cardano/Ledger/Plutus/ToPlutusData.hs:
```diff
@@ -0,0 +1,259 @@
+
+instance ToPlutusData Word32 where
+  toPlutusData w32 = I (toInteger @Word32 w32)
+  fromPlutusData (I n) | n >= 0 = Just $ fromInteger @Word32 n
```
Same should apply to other `Word` like conversions
```suggestion
  fromPlutusData (I n) | n >= 0 && n <= toInteger (maxBound @Word32) =
    Just $ fromInteger @Word32 n
```

lehins commented on libs/cardano-ledger-core/src/Cardano/Ledger/Plutus/ToPlutusData.hs:
```diff
@@ -0,0 +1,259 @@
+  toPlutusData (SJust x) = Constr 1 [toPlutusData x]
+  fromPlutusData (Constr 0 []) = Just SNothing
+  fromPlutusData (Constr 1 [x]) = SJust <$> fromPlutusData x
+  fromPlutusData _ = Nothing
```
We don't need to translate that either. In the matter of fact, because this instance is available the implementation of `pparamUpdateFromData` is incorrect, because it assumes pparam values are encoded as StrictMaybe
```suggestion
```

TimSheard replied:
I am not sure I understand this.

lehins replied:
You already fixed the problem that I was talking about. So you are no longer even using this instance, so it can be safely removed.

lehins replied:
This instance is not needed and should be removed

lehins commented on libs/cardano-ledger-core/src/Cardano/Ledger/Plutus/ToPlutusData.hs:
```diff
@@ -0,0 +1,259 @@
+              Nothing -> ppu
+              Just t -> ppu & lense .~ t
+            else ppu
+pparamUpdateFromData _ _ = Nothing
```
That's just a monadic fold, also we need to guard against duplicates and make sure we are not expecting StrictMaybe for values
```suggestion
pparamUpdateFromData pparamMap (Map pairs) = 
  foldlM accum (emptyPParamsUpdate, pparamMap) pairs
  where
    accum (!ppu, !leftOverMap) (key, value) = do
      tag <- fromPlutusData key
      val <- fromPlutusData value
      PParam _ ppuL <- Map.lookup tag leftOverMap
      pure (ppu $ ppuL .~ SJust val, Map.delete tag leftOverMap)
```

lehins replied:
I don't have a type checker in github :grin: 

But is better if I make a half working suggestion than try to explain in words what I mean.

TimSheard replied:
Took a while to debug this.
The biggest problem was the order of the pattern matching in the do

lehins commented on libs/cardano-ledger-core/src/Cardano/Ledger/Plutus/ToPlutusData.hs:
```diff
@@ -0,0 +1,259 @@
+makePParamMap :: [PParam era] -> Map.Map Integer (PParam era)
+makePParamMap xs = foldl' accum Map.empty xs
+  where
+    accum m p@(PParam n _) = Map.insert n p m
```
This will be quite a bit faster
```suggestion
makePParamMap xs = Map.fromList [(n, p) | p@(PParam n _) <- xs]
```

lehins commented on eras/conway/impl/src/Cardano/Ledger/Conway/PParamsUpdate.hs:
```diff
@@ -0,0 +1,127 @@
+
+-- | A Decoder for the Conway era
+decodeConway :: ConwayEraPParams era => Decoder s (PParamsUpdate era)
+decodeConway = decodePParamsUpdate conwayPParamMap
```
Please remove that as well, we'll do this at a later time.
```suggestion
```

lehins reviewed (changes requested):
There are some issues in the PR, but the overall direction is great!

The only thing that is missing now is the actual translation to PlutusV3 context and roundtrip tests that should have the reusable functionality placed in `cardano-ledger-core:testlib:Test.Cardano.Ledger.Plutus.ToPlutusData`
and the actual Conway tests should go into `cardano-ledger-conway:tests:Test.Cardano.Ledger.Conway.Plutus.ToPlutusData`

lehins commented on eras/conway/impl/src/Cardano/Ledger/Conway/PParams.hs:
```diff
@@ -1295,3 +1301,105 @@ conwayModifiedPPGroups
+
+instance ToPlutusData CoinPerByte where
+  toPlutusData (CoinPerByte c) = toPlutusData @Coin c
+  fromPlutusData x = CoinPerByte <$> fromPlutusData @Coin x
```
This can be derived using GeneralizedNewtypeDeriving

TimSheard replied:
done

lehins commented on eras/conway/impl/src/Cardano/Ledger/Conway/PParams.hs:
```diff
@@ -1295,3 +1301,105 @@ conwayModifiedPPGroups
+  fromPlutusData = pparamUpdateFromData conwayPParamMap
+
+class ConwayEraPlutusTxInfo era where
+  ppuToData :: PParamsUpdate era -> Data
```
This class should go into a new module `Cardano.Ledger.Conway.Plutus.Context`

And the name of the function should be `toPlutusChangedParameters`

It also needs to be parameterized on language

Also note the change in the return type.

```suggestion
class EraPlutusTxInfo l era => ConwayEraPlutusTxInfo (l :: Language) era where
  toPlutusChangedParameters :: proxy l -> PParamsUpdate era -> PV3.ChangedParameters
```

lehins commented on libs/cardano-ledger-core/testlib/Test/Cardano/Ledger/Plutus/ToPlutusData.hs:
```diff
@@ -0,0 +1,72 @@
+plutusDataRoundtripSpec =
+  do
+    describe "roundtrip ToPlutusData" $
+      do
```
Slightly unusually formatting. Also this is not a reusable test case, so it should just go into the test suite in to `Test.Cardano.Ledger.PlutusSpec` module, instead of be exported from `testlib`
```suggestion
plutusDataRoundtripSpec = do
  describe "roundtrip ToPlutusData" $ do
```

lehins commented on libs/cardano-ledger-core/testlib/Test/Cardano/Ledger/Plutus/ToPlutusData.hs:
```diff
@@ -0,0 +1,72 @@
+              ("\nIntermediate Data\n" ++ show (toPlutusData y) ++ "\n")
+              (Just (y :: x) === fromPlutusData (toPlutusData y))
+        )
+    )
```
hspec will already print the generated value during test failure, so there is no need to print it again. Also using `shouldBe` will give all of the information necessary during the test failure (it'll even diff it). That is why `hspec` is great: 
```suggestion
  prop (show (typeRep @x)) $ \ y ->
    fromPlutusData (toPlutusData y) `shouldBe` Just (y :: x)
```

lehins commented on libs/cardano-ledger-core/testlib/Test/Cardano/Ledger/Plutus/ToPlutusData.hs:
```diff
@@ -0,0 +1,72 @@
+{-# LANGUAGE TypeApplications #-}
+
+module Test.Cardano.Ledger.Plutus.ToPlutusData (
+  roundtrip,
```
We have bunch of round tripping properties in this package. It'll be good to distinguish with a suffix. Also adding suffix `Spec` will let the user know what it is. It could have been either a `Property` or `Expectation`, but it is a `Spec`, which is totally fine, just a nice convention to distinguish between the three concepts.
```suggestion
  roundTripPlutusDataSpec,
```

lehins commented on libs/cardano-ledger-core/src/Cardano/Ledger/Plutus/ToPlutusData.hs:
```diff
@@ -0,0 +1,139 @@
+instance ToPlutusData EpochInterval where
+  toPlutusData (EpochInterval w32) = I (toInteger @Word32 w32)
+  fromPlutusData (I n) | n >= 0 = Just . EpochInterval $ fromInteger @Word32 n
+  fromPlutusData _ = Nothing
```
It is better to derive this. This way we guarantee that the upper bound check is also done:
```haskell
    | n >= 0 && n <= toInteger (maxBound @Word32) =
```
```suggestion
deriving instance ToPlutusData EpochInterval
```

lehins commented on libs/cardano-ledger-core/src/Cardano/Ledger/Plutus/ToPlutusData.hs:
```diff
@@ -0,0 +1,139 @@
+instance ToPlutusData Coin where
+  toPlutusData (Coin n) = I n
+  fromPlutusData (I n) = Just (Coin n)
+  fromPlutusData _ = Nothing
```
```suggestion
deriving instance ToPlutusData Coin
```

lehins commented on libs/cardano-ledger-core/testlib/Test/Cardano/Ledger/Plutus/ToPlutusData.hs:
```diff
@@ -0,0 +1,72 @@
+
+-- | Test the roundtrip property via 'fromPlutusData' and 'toPlutusData' for any type
+--   meeting all the constraints below.
+roundtrip :: forall x. (Typeable x, Show x, Eq x, Arbitrary x, ToPlutusData x) => Spec
```
It is important to add `HasCallStack` to this kind of helpers:
```suggestion
roundtrip :: forall x. (HasCallStack, Typeable x, Show x, Eq x, Arbitrary x, ToPlutusData x) => Spec
```

lehins commented on eras/conway/impl/src/Cardano/Ledger/Conway/PParams.hs:
```diff
@@ -113,6 +114,12 @@ import GHC.Generics (Generic)
 import Numeric.Natural (Natural)
+import PlutusLedgerApi.Common (Data (..))
+
+-- import Cardano.Ledger.Conway.Plutus.ToPlutusData(conwayPParamMap)
```
dead code
```suggestion
```

lehins commented on eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/Plutus/ToPlutusData.hs:
```diff
@@ -0,0 +1,26 @@
+import Test.Cardano.Ledger.Core.Arbitrary ()
+import Test.Cardano.Ledger.Plutus.ToPlutusData (roundtrip)
+
+import Test.Hspec
```
Common re-exports all of the stuff necessary for testing.
```suggestion
import Test.Cardano.Ledger.Common
```

lehins commented on eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/Plutus/ToPlutusData.hs:
```diff
@@ -0,0 +1,26 @@
+{-# LANGUAGE TypeApplications #-}
+
+module Test.Cardano.Ledger.Conway.Plutus.ToPlutusData where
```
There are no reusable test here either, so they should got into the test suite under this module name, rather than in the `testlib`.

For this also the convention with hspec tests is to export only a single function `spec` that has all of the tests in it. 
```suggestion
module Test.Cardano.Ledger.Conway.PlutusSpec (spec) where
```

lehins commented on eras/conway/impl/test/Main.hs:
```diff
@@ -34,3 +35,4 @@ main =
         CostModelsSpec.spec @Conway
       describe "TxWits" $ do
         TxWitsSpec.spec @Conway
+      conwayPlutusDataRoundtripSpec
```
So, this will look like so:
```suggestion
      PlutusSpec.spec
```

lehins commented on eras/conway/impl/test/Main.hs:
```diff
@@ -13,6 +13,7 @@ import qualified Test.Cardano.Ledger.Conway.DRepRatifySpec as DRepRatify
 import qualified Test.Cardano.Ledger.Conway.GenesisSpec as Genesis
 import qualified Test.Cardano.Ledger.Conway.GovActionReorderSpec as GovActionReorder
 import qualified Test.Cardano.Ledger.Conway.Imp as ConwayImp
+import Test.Cardano.Ledger.Conway.Plutus.ToPlutusData (conwayPlutusDataRoundtripSpec)
```
See my comments below about this.
```suggestion
import Test.Cardano.Ledger.Conway.PlutusSpec as PlutsuSpec
```

lehins commented on eras/conway/impl/src/Cardano/Ledger/Conway.hs:
```diff
@@ -26,7 +26,8 @@ import Cardano.Ledger.Conway.Rules ()
 import Cardano.Ledger.Conway.Tx ()
-import Cardano.Ledger.Conway.TxInfo ()
+
+-- import Cardano.Ledger.Conway.TxInfo ()
```
This import is necessary. It brings orphan instances into scope toegther with the ConwayEra type in order to un-orphan them.
```suggestion
import Cardano.Ledger.Conway.TxInfo ()
```

lehins commented on eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/Plutus/ToPlutusData.hs:
```diff
@@ -0,0 +1,26 @@
+conwayPlutusDataRoundtripSpec =
+  do
+    describe "roundtrip ToPlutusData Conway instances" $
+      do
```
Inconsistent placement of `do` and rename to `spec`
```suggestion
spec :: Spec
spec = do
  describe "roundtrip ToPlutusData Conway instances" $ do
```

lehins commented on libs/cardano-ledger-core/src/Cardano/Ledger/Core/PParams.hs:
```diff
@@ -546,3 +559,46 @@ downgradePParamsUpdate ::
+      PParam _ ppuL <- Map.lookup tag leftOverMap
+      val <- fromPlutusData value
+      pure (ppu & ppuL .~ SJust val, Map.delete tag leftOverMap)
+pparamUpdateFromData _ _ = Nothing
```
It would be better to put these two functions into a new module `Cardano.Ledger.Conway.Plutus.Context` together with `ConwayEraPlutusTxInfo`, because those functions will be used for creating instances for that class. This way it will be easier for people to find.

This functionality is very much plutus specific and can't be used prior to Conway, so we don't need to to keep it in the `Cardano.Ledger.Core.PParams`

The `PParam` GADT, on the other hand, we'll be able to use other places later on, so this is a perfect home for it.

TimSheard replied:
done

lehins reviewed (commented):
This looks great. Semantically that is exactly what we need.
Just a few comments on shuffling things around for consistency with the rest of the codebase

lehins committed:
Switch away from using AllowAmbiguousTypes in favor or proxy

This change is needed for consistency with how the rest of the
translation works and to avoid constructing unnecessary `Proxy`,
since `Plutus a` is already used as a proxy.

lehins committed:
Make `ConwayEraPlutusTxInfo` for `ConwayEra` specific to `PlutusV3`

This is necessary to remove such instance for PlutusV1, PlutusV2 and the
future languages past PlutusV3.

Make `EraPlutusTxInfo` usperclass of `ConwayEraPlutusTxInfo` for
consistent hierarchy of classes

lehins committed:
Move test cases int the test suites from `testlib`s

lehins reviewed (approved):
@TimSheard There was a few things still needed to be rearranged in this PR, but because we will try to make a release today and this functionality is needed for the next release I took upon myself to finish this up. Check out the last few commits to see what I've done.
